### PR TITLE
Refactor type cache with templates

### PR DESCRIPTION
**Pull request by krishnangovindraj**
## Usage and product changes
Remove duplication in type_cache using generics.

## Implementation
* Introduces the `KindAPI` to hold properties of the vertex types (`EntityType, AttributeType, RelationType, RoleType`) 
* Refactor all `get_X_type_y(x)` methods to be `get_y<X>(x: X)`  where X is a fundamental type, and y is some property of it such as supertype, label, etc.
* Create a `selection` submodule Introducing the traits needed to do this
  - CacheGetter:  To pick the member cache of TypeCache corresponding to  X
  - HasCommonCache:  HasCommonCache trait - to restrict the get_X_type_y methods to X which implement KindAPI
  - HasOwnerPlayerCache:  to restrict the get_X_type_y methods to X which implement OwnerAPI and PlayerAPI
* Move functions for construction of TypeCache & all subcaches into a `construction` submodule.

### CODE REVIEW

File: concept/type_/type_cache.rs [hunk 1]
```diff
@@ -608,6 +608,55 @@ impl TypeCache {
...
+        // TODO: All the unwrap
+        Some(T::select(self, type_).supertype.as_ref()?.clone())
+    }
+
+    pub(crate) fn __generalised__get_label<'a, T>(&self, type_: T) -> &Label<'static>
+    where T: TypeAPI<'a> + CommonTypeCacheSelector,
+          <T as CommonTypeCacheSelector>::CacheParameter: 'static
+    {
+        &T::select(self, type_).label
+    }
```
Comment (krishnangovindraj) [inline concept/type_/type_cache.rs:626]: These can replace the code above that's repeated for each of EntityType, RelationType, AttributeType, RoleType

File: concept/type_/type_cache.rs [hunk 2]
```diff
@@ -608,6 +608,55 @@ impl TypeCache {
...
+    fn select<'cache>(type_cache: &'cache TypeCache, type_: RelationType<'a>) -> &'cache CommonTypeCache<Self::CacheParameter> {
+        &TypeCache::get_relation_type_cache(&type_cache.relation_types, type_.vertex()).unwrap().common_type_cache
+    }
+}
+
+impl<'a> CommonTypeCacheSelector for RoleType<'a> {
+    type CacheParameter = RoleType<'static>;
+    fn select<'cache>(type_cache: &'cache TypeCache, type_: RoleType<'a>) -> &'cache CommonTypeCache<Self::CacheParameter> {
+        &TypeCache::get_role_type_cache(&type_cache.role_types, type_.vertex()).unwrap().type_api_cache_
+    }
```
Comment (krishnangovindraj) [inline concept/type_/type_cache.rs:659]: The equivalent of `ReadableType` created for type_reader

File: concept/type_/type_manager.rs [hunk 3]
```diff
@@ -308,10 +308,10 @@ impl<Snapshot: ReadableSnapshot> TypeManager<Snapshot>
...
 
     get_type_label_methods! {
-        fn get_entity_type_label() -> EntityType = get_entity_type_label;
-        fn get_relation_type_label() -> RelationType = get_relation_type_label;
-        fn get_role_type_label() -> RoleType = get_role_type_label;
-        fn get_attribute_type_label() -> AttributeType = get_attribute_type_label;
+        fn get_entity_type_label() -> EntityType = __generalised__get_label;
+        fn get_relation_type_label() -> RelationType = __generalised__get_label;
+        fn get_role_type_label() -> RoleType = __generalised__get_label;
+        fn get_attribute_type_label() -> AttributeType = __generalised__get_label;
```
Comment (krishnangovindraj) [inline concept/type_/type_manager.rs:314]: `test_type.rs` runs fine with these changed.

**Review by krishnangovindraj — COMMENTED** [review]
Self review

File: concept/type_/type_cache.rs [hunk 4]
```diff
@@ -608,6 +608,55 @@ impl TypeCache {
...
         attribute_type_caches[as_u16 as usize].as_ref()
     }
+
+
+    pub(crate) fn __generalised__get_supertype<'a, T>(&self, type_: T) -> Option<T::CacheParameter>
+    where T: TypeAPI<'a> + CommonTypeCacheSelector
+    {
+        // TODO: Why does this not return &Option<EntityType<'static>> ?
+        // TODO: All the unwrap
+        Some(T::select(self, type_).supertype.as_ref()?.clone())
```
Comment (flyingsilverfin) [inline concept/type_/type_cache.rs:618]: `get_cache` on a `CommonTypeCacheGetter`

**Pull request merged**